MdeModulePkg/Core/Dxe/Mem/Page.c

```diff
@@ -1765,6 +1772,157 @@ CoreFreePages (
+
+**/
+EFI_STATUS
+EFIAPI
+CoreFreePages (
+  IN EFI_PHYSICAL_ADDRESS  Memory,
+  IN UINTN                 NumberOfPages
+  )
+{
+  EFI_STATUS                       Status;
+  EFI_MEMORY_TYPE                  MemoryType;
+  UINTN                            Alignment;
+  BOOLEAN                          IsGuarded;
+  UINTN                            NumberOfPagesAligned;
+  EFI_GCD_MEMORY_SPACE_DESCRIPTOR  Descriptor;
+  UINT64                           CurrentAddress;
+  UINT64                           EndAddress;
+  UINT64                           DescEnd;
+  UINT64                           CurrentLength;
+  EFI_STATUS                       OperationStatus;
+
+  NumberOfPagesAligned = NumberOfPages;
+  OperationStatus      = EFI_SUCCESS;
+
+  if (NumberOfPages == 0) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  CoreAcquireMemoryLock ();
+  Status = CoreFindEntryToFree (
+             Memory,
+             &NumberOfPagesAligned,
+             &IsGuarded,
+             &Alignment,
+             &MemoryType
+             );
+
+  // now we drop the lock to avoid deadlocks when debug clearing memory, this is safe to do because even if an
+  // interrupt came in and changed the memory map, it would be a bug if it was for this region (since it is being
+  // freed) and the only reason we took the lock now was to parse the list, which we are done with for now)
+  CoreReleaseMemoryLock ();
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  // The free pages code will attempt to zero the freed pages if DebugClearMemory is enabled. However, there is no
+  // guarantee that the memory is writeable, which can cause the core to crash. Furthermore, if DebugClearMemory is
+  // not enabled and NX memory protection is not enabled, we can end up passing pages freed here directly to another
+  // caller in the next allocate pages call, which may not be writeable, which is the expection. So, we need to ensure
+  // the memory is writeable before we attempt to zero it or pass it out. We must do this outside of the memory lock to
+  // avoid deadlocks going through the GCD, which may cause page allocations to occur. It is safe to call the GCD from
+  // here, because the GCD only uses pool backed memory and CoreFreePoolPages() does not call through this path, so we
+  // won't hit a deadlock on the GCD lock. We don't attempt to call this before gCpu is initialized, because before
+  // that point, page attributes can't be set anyway and we may be very early in GCD init and not have the GCD
+  // populated yet.
+  if (gCpu == NULL) {
+    return CoreInternalFreePages (Memory, NumberOfPagesAligned, MemoryType, IsGuarded);
+  }
+
+  CurrentAddress = Memory;
+  EndAddress     = Memory + EFI_PAGES_TO_SIZE (NumberOfPagesAligned);
+  while (CurrentAddress < EndAddress) {
+    Status = CoreGetMemorySpaceDescriptor (
+               CurrentAddress,
+               &Descriptor
+               );
+
+    if (EFI_ERROR (Status)) {
+      DEBUG ((

```

Comment:
Is this really an error condition and an ASSERT() condition?  Since it will continue in release builds, should this be a WARN message with no ASSERT() so the log will should some memory pages were not actually freed?

Reply:
I’m fine to change this, my rationale is I do consider it an error if a caller passed in a memory address to free that didn’t exist in the GCD, that means the caller either messed up the pointer it had to point to some place in memory that doesn’t exist or the core incorrectly gave out some pointer that doesn’t exist, in either case the existence of pointers to non-existent memory is an error case to me that I would want to catch with an assert. In release builds, you would see the error message and know some pages were not freed (same with if it were a warn).

But, if you prefer warn and no assert, I can switch to that.